File: core/src/main/java/pcb/uwu/core/UnitCounter.java
package pcb.uwu.core;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.function.Function;

import static java.util.Collections.emptyMap;
import static java.util.stream.Collectors.toSet;

public class UnitCounter {

	private final Map<Class<? extends BaseUnit>, Map<BaseUnit, Integer>> powers;

	private static final char NEGATIVE = '⁻';
	private static final char[] POWERS = new char[] {'⁰', '¹', '²', '³', '⁴', '⁵', '⁶', '⁷', '⁸', '⁹'};

	// region constructors

	public UnitCounter() {
		powers = new HashMap<>();
	}

	public UnitCounter(BaseUnit canonic) {
		this();

		addPower(canonic.getBaseUnitType(), canonic, 1);
	}

	private UnitCounter(UnitCounter source) {
		this(source.powers);
	}

	private UnitCounter(Map<Class<? extends BaseUnit>, Map<BaseUnit, Integer>> powers) {
		this.powers = new HashMap<>(powers);
	}

	// endregion

	public UnitCounter major(UnitCounter unitCounter) {
		UnitCounter result = new UnitCounter(this);

		unitCounter.powers.forEach(result::addMajor);

		return result;
	}

	public UnitCounter major(BaseUnit unit) {
		return major(unit, 1);
	}

	public UnitCounter major(BaseUnit unit, int counts) {
		UnitCounter result = new UnitCounter(this);

		result.addPower(unit.getBaseUnitType(), unit, counts);

		return result;
	}

	public UnitCounter minor(UnitCounter unitCounter) {
		UnitCounter result = new UnitCounter(this);

		unitCounter.powers.forEach(result::addMinor);

		return result;
	}

	public UnitCounter minor(BaseUnit unit) {
		return minor(unit, 1);
	}

	public UnitCounter minor(BaseUnit unit, int counts) {
		UnitCounter result = new UnitCounter(this);

		result.addPower(unit.getBaseUnitType(), unit, -counts);

		return result;
	}

	public int get(BaseUnit unit) {
		return powers.getOrDefault(unit.getBaseUnitType(), emptyMap()).getOrDefault(unit, 0);
	}

	public Set<BaseUnit> getBaseUnits() {
		return powers.values().stream()
				.map(Map::keySet)
				.flatMap(Collection::stream)
				.collect(toSet());
	}

	public String asString(Function<Unit, String> majorString, Function<Unit, String> minorString, String separator) {
		String result = "";

		Set<BaseUnit> major = powers.values().stream()
				.map(Map::entrySet)
				.flatMap(Collection::stream)
				.filter(entry -> entry.getValue() > 0)
				.map(Entry::getKey)
				.collect(toSet());

		boolean first = true;

		for (BaseUnit unit : major) {
			String power = buildPower(get(unit));
			if (first) {
				result += majorString.apply(unit) + power;
			} else {
				result += minorString.apply(unit) + power;
			}
			first = false;
		}

		Set<BaseUnit> minor = powers.values().stream()
				.map(Map::entrySet)
				.flatMap(Collection::stream)
				.filter(entry -> entry.getValue() < 0)
				.map(Entry::getKey)
				.collect(toSet());

		if (!minor.isEmpty()) {
			result += separator;
		}

		for (BaseUnit unit : minor) {
			String power = buildPower(Math.abs(get(unit)));
			result += minorString.apply(unit) + power;
		}

		return result;
	}

	// region private methods

	private String buildPower(int power) {
		String result = "";
		boolean negative = power < 0;

		// Omit neutral power of 1
		if (power == 1) {
			return result;
		}

		if (power == 0) {
			return String.valueOf(POWERS[0]);
		}

		power = Math.abs(power);

		for (;power > 0 ; power /= 10)
		{
			result = POWERS[power % 10] + result;
		}

		if (negative) {
			return NEGATIVE + result;
		}

		return result;
	}

	private void addMajor(Class<? extends BaseUnit> clazz, Map<BaseUnit, Integer> unitCounts) {
		for (Entry<BaseUnit, Integer> entry : unitCounts.entrySet()) {
			BaseUnit unit = entry.getKey();
			Integer count = entry.getValue();

			addPower(clazz, unit, count);
		}
	}

	private void addMinor(Class<? extends BaseUnit> clazz, Map<BaseUnit, Integer> unitCounts) {
		for (Entry<BaseUnit, Integer> entry : unitCounts.entrySet()) {
			BaseUnit unit = entry.getKey();
			Integer count = entry.getValue();

			addPower(clazz, unit, -count);
		}
	}

	private void addPower(Class<? extends BaseUnit> clazz, BaseUnit unit, Integer count) {
		if (count == 0) return;

		Map<BaseUnit, Integer> internal = powers.getOrDefault(clazz, new HashMap<>());
		int result = internal.getOrDefault(unit, 0) + count;

		if (result == 0) {
			internal.remove(unit);

			if (internal.isEmpty()) {
				powers.remove(clazz);
			}
		} else {
			internal.put(unit, result);

			powers.put(clazz, internal);
		}
	}

	public boolean isEmpty() {
		return powers.isEmpty();
	}

	@SuppressWarnings("unchecked")
	public <U extends BaseUnit> U findUnit(Class<U> unitClass) {
		return getUnit(unitClass, getBaseUnits()); //todo drop after normalisation is applied
	}

	@SuppressWarnings("unchecked")
	private <U extends BaseUnit> U getUnit(Class<U> unitClass, Set<BaseUnit> units) {
		for (Unit unit : units) {
			if (unitClass.isAssignableFrom(unit.getClass())) return (U) unit;
		}

		return null;
	}

	// endregion
}

File: core/src/main/java/pcb/uwu/core/UnitCounter.kt
package pcb.uwu.core

import pcb.uwu.core.UnitCounter.UnitCount.Companion.EMPTY_BASE_UNIT_COUNT
import pcb.uwu.exception.OffendingUnitException
import pcb.uwu.unit.base.ScalarUnit
import java.util.Objects
import java.util.function.Function
import kotlin.math.abs

class UnitCounter
{
    class UnitCount(val unit: BaseUnit, val count: Int)
    {
        @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
        override fun equals(that: Any?): Boolean
        {
            if (that is UnitCount)
            {
                return unit == that.unit &&
                        count == that.count
            }
            return false
        }

        override fun hashCode(): Int
        {
            return Objects.hash(unit, count)
        }

        companion object
        {
            val EMPTY_BASE_UNIT_COUNT = UnitCount(ScalarUnit.SCALAR, 0)
        }
    }

    // region private fields

    private val counts: MutableMap<Class<out BaseUnit>, UnitCount>

    // endregion

    // region constructors

    constructor() : this(HashMap())

    constructor(unit: BaseUnit) : this()
    {
        addPower(unit.baseUnitType, unit, 1)
    }

    constructor(source: UnitCounter) : this(source.counts)

    private constructor(counts: Map<Class<out BaseUnit>, UnitCount>)
    {
        this.counts = HashMap(counts)
    }

    // endregion

    fun invert(): UnitCounter
    {
        val result: MutableMap<Class<out BaseUnit>, UnitCount> = HashMap()

        counts.forEach { (clazz, unitCount) ->
            result[clazz] = UnitCount(unitCount.unit, -unitCount.count)
        }

        return UnitCounter(result)
    }

    fun major(unitCounter: UnitCounter): UnitCounter
    {
        val result = UnitCounter(this)
        unitCounter.counts.forEach { (clazz, unitCount) ->
            result.addMajor(clazz, unitCount)
        }
        return result
    }

    @JvmOverloads
    fun major(unit: BaseUnit, counts: Int = 1): UnitCounter
    {
        val result = UnitCounter(this)
        result.addPower(unit.baseUnitType, unit, counts)
        return result
    }

    fun minor(unitCounter: UnitCounter): UnitCounter
    {
        val result = UnitCounter(this)
        unitCounter.counts.forEach { (clazz, unitCount) ->
            result.addMinor(clazz, unitCount)
        }
        return result
    }

    @JvmOverloads
    fun minor(unit: BaseUnit, counts: Int = 1): UnitCounter
    {
        val result = UnitCounter(this)
        result.addPower(unit.baseUnitType, unit, -counts)
        return result
    }

    operator fun get(unit: BaseUnit) =
        counts[unit.baseUnitType] ?: EMPTY_BASE_UNIT_COUNT

    val baseUnits: Collection<UnitCount>
        get() = counts.values

    fun asString(majorString: Function<Unit, String>,
                 minorString: Function<Unit, String>)
            : String
    {
        val result = StringBuilder()
        val baseUnits = counts.values.asSequence()
            .map { it.unit }
            .sorted()
            .toList()

        var first = true
        for (unit in baseUnits)
        {
            val power = buildPower(get(unit).count)
            if (first)
            {
                result.append(majorString.apply(unit)).append(power)
            }
            else
            {
                result.append('⋅').append(minorString.apply(unit)).append(power)
            }
            first = false
        }
        return result.toString()
    }

    // region private methods

    private fun buildPower(power: Int): String
    {
        var mutablePower = power
        val result = StringBuilder()
        val negative = mutablePower < 0

        // Omit neutral power of 1
        if (mutablePower == 1)
        {
            return result.toString()
        }
        if (mutablePower == 0)
        {
            return POWERS[0].toString()
        }
        mutablePower = abs(mutablePower)
        while (mutablePower > 0)
        {
            result.insert(0, POWERS[mutablePower % 10])
            mutablePower /= 10
        }
        return if (negative)
        {
            "$NEGATIVE$result"
        }
        else result.toString()
    }

    private fun addMajor(clazz: Class<out BaseUnit>, unitCount: UnitCount)
    {
        addPower(clazz, unitCount.unit, unitCount.count)
    }

    private fun addMinor(clazz: Class<out BaseUnit>, unitCount: UnitCount)
    {
        addPower(clazz, unitCount.unit, -unitCount.count)
    }

    private fun addPower(clazz: Class<out BaseUnit>, unit: BaseUnit, count: Int)
    {
        if (count == 0) return
        val unitCount = counts[clazz]
        val result = if (unitCount == null)
        {
            count
        }
        else
        {
            if (unitCount.unit != unit)
            {
                throw OffendingUnitException("${unitCount.unit.singularName} and ${unit.singularName} are both ${unit.baseUnitType.simpleName}")
            }
            unitCount.count + count
        }
        if (result == 0)
        {
            counts.remove(clazz)
        }
        else
        {
            counts[clazz] = UnitCount(unit, result)
        }
    }

    fun <U : BaseUnit> findUnit(unitClass: Class<U>): U?
    {
        return getUnit(unitClass, baseUnits)
    }

    @Suppress("UNCHECKED_CAST")
    private fun <U : BaseUnit> getUnit(unitClass: Class<U>, units: Collection<UnitCount>): U?
    {
        for (unitCount in units)
        {
            val unit = unitCount.unit
            if (unitClass.isAssignableFrom(unit.javaClass)) return unit as U
        }
        return null
    }

    val isEmpty: Boolean
        get() = counts.isEmpty()

    // endregion

    // region Object

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun equals(that: Any?): Boolean
    {
        if (that is UnitCounter)
        {
            return this.counts == that.counts
        }
        return false
    }

    override fun hashCode(): Int
    {
        return counts.hashCode()
    }

    // endregion

    companion object
    {
        private const val NEGATIVE = '⁻'
        private val POWERS = charArrayOf('⁰', '¹', '²', '³', '⁴', '⁵', '⁶', '⁷', '⁸', '⁹')
    }
}